### ISSUE TITLE
Pi-Hole does not automatically provide DNS to network devices

### ISSUE THREAD

**Issue by Ferkner** [pi-hole/pi-hole#997]
I installed pi-hole on my Raspberry Pi 2, and it does not work as advertised. Unless I manually go onto each machine and change the DNS, pi-hole does not work, and the whole appeal of it is that it does this for you automatically for everything on your network.

I gave the Pi a static IP (192.168.2.222) outside of pi-hole. I set up pihole with that IP. I went it my modem/router manually set the DNS as 192.138.2.222. The modem is a Bell Home Hub 2000 (Sagemcom 5250) modem and router combo. It requires a secondary DNS, so I entered 192.168.2.222 for both. IT's as though the router is not paying attention to my manually inputed DNS. The only computer it actually works on is my desktop, and I connect to the outside world though the PIA VPN. All the computers are on wifi, including my desktop with VPN.

When I connect to the VPN, it creates a virtual Ethernet 2 connection. The icon in my task bar changes from the arches signifying wifi to a computer with an Ethernet plug. The Pi is connected to the network by Ethernet. When I set it up, the only option was eth0 when asked what it will "listen" to. Is this why it isn't automatically working for my other devices? My desktop is fake-connected to Ethernet because of my VPN, and when it is I can see my DNS Server is 192.168.2.222 when I do ipconfig in the command prompt. When I disconnect from the VPN, pi-hole stops working.

So if this is the problem, is there a way for pi-hole to listen on both eth0 and wlan0 at the same time, so that anything that connects to my network goes through pi-hole?

I know I can set the DNS for each computer to get it to work, but that defeats the purpose of pi-hole for me. And I cannot change the DNS on a couple of the devices we have.

**In raising this issue, I confirm the following (please check boxes, eg [X]) Failure to fill the template will close your issue:**

- [X] I have read and understood the [contributors guide](https://github.com/pi-hole/pi-hole/blob/master/CONTRIBUTING.md).
- [X] The issue I am reporting can be *replicated*
- [X] The issue I'm reporting isn't a duplicate (see [FAQs](https://github.com/pi-hole/pi-hole/wiki/FAQs), [closed issues](https://github.com/pi-hole/pi-hole/issues?utf8=%E2%9C%93&q=is%3Aissue%20is%3Aclosed%20), and [open issues](https://github.com/pi-hole/pi-hole/issues)).

**How familiar are you with the codebase?:**

- [] 1 (very unfamiliar)
- [X] 2
- [] 3
- [] 4
- [] 5
- [] 6
- [] 7
- [] 8
- [] 9
- [] 10 (very familiar)

---
**[FEATURE REQUEST | QUESTION | OTHER]:**

_{replace this section with your content or delete if not a FEATURE REQUEST/QUESTION/OTHER}_

**[BUG | ISSUE] Expected Behaviour:**

**[BUG | ISSUE] Actual Behaviour:**

**[BUG | ISSUE] Steps to reproduce:**

-
-
-
-

**(Optional) Debug Log generated by `pihole -d`:**

`http://termbin.com/<something>`

_This template was created based on the work of [`udemy-dl`](https://github.com/nishad/udemy-dl/blob/master/LICENSE)._

**Comment by Tntdruid** [comment]
Install dhcp server on the pi

**Comment by DL6ER** [comment]
I agree with what @Tntdruid suggested. In the next version you will be able to enable an integrated DHCP server with one click. We implemented this because surprisingly many routers doe not allow to change the DNS settings at all or they don't communicate the new DNS server to the connected devices properly. I guess it is not necessary to ask this but I do it nevertheless: have to reconnected the devices after changing the settings in the router? New networking configuration is only transferred at connection time, not while already connected.

> is there a way for pi-hole to listen on both eth0 and wlan0 at the same time

Yes, but that is currently only possible by editing the file `/etc/dnsmasq.d/01-pihole.conf`. Remove the line `interface=eth0` to have it listening on all networking devices at the same time.

**Comment by Ferkner** [comment]
I tried installing a DCHP server last night as well. That did not seem to do anything by break to connection between my computer when the VPN was turned off and my network (as well as my phone). Again, if pi-hole was only listening to ethernet connections, my devices would not be intercepted by pi-hole and would just connect directly to the Home Hub router which had the DCHP server disabled.

I tried editing the `/etc/dnsmasq.d/01-pihole.conf` file last night, but the instructions I used said to add wlan to the interface line. That didn't work, obviously. But I will try removing the line you mentioned and give it another go. I assume I can just comment it out instead of completely deleting it (I'm still fairly new to playing around with the Raspberry Pi).

And yes, I reconnected my devices after changing the router settings. I also made the devices forget the connection so that they would make a fresh connection. I also flushed the DNS on my tablet, and rebooted the router.

**Comment by DL6ER** [comment]
>  I assume I can just comment it out instead of completely deleting it (I'm still fairly new to playing around with the Raspberry Pi).

Sure, you can also put a `#` in front of the line.

Can you maybe draw a little picture so we understand better? I'm still not sure how your VPN is working in your configuration - do you connect to somewhere else with it? 

Also I don't understand
> my devices would not be intercepted by pi-hole and would just connect directly to the Home Hub router which had the DCHP server disabled

If DHCP in the router is disabled they should not be able to see the router at all, i.e. they shouldn't have internet access unless they get managed by the DHCP server on your Raspberry. If still everything is working when you switch off the DHCP server in the router, then there has to be something different somewhere. Maybe the VPN is causing problems, but again, I don't know in which context it is included in your network.

**Comment by Ferkner** [comment]
> If DHCP in the router is disabled they should not be able to see the router at all, i.e. they shouldn't have internet access unless they get managed by the DHCP server on your Raspberry. If still everything is working when you switch off the DHCP server in the router, then there has to be something different somewhere. Maybe the VPN is causing problems, but again, I don't know in which context it is included in your network.

With DCHP disabled on my router, my devices connect to the router but have no internet access. All of them except my desktop when connected to the VPN.

My VPN is installed locally on my desktop computer. Initially when I start my computer, it will connect to my router with WiFi as normal. Once my VPN program starts (in this case, Private Internet Access), it takes over my connection and creates an adapter called Ethernet 2 to connect the computer to my network and to the outside world. It also has an Internet kill switch, so that if it stops working for some reason, it will kill my Internet connection until I manually reconnect.

I still have access to my home network and anything else that lives on it. I am thinking that because the VPN creates that Ethernet 2 adapter, the pi-hole sees it. When I do ipconfig on my computer, it spits back the IP that I assigned to the Pi as my DNS server. If I turn my VPN off and restore my regular wireless connection on my desktop, pi-hole no longer works. If I remember correctly, the VPN will create it owns DCHP and DNS when pi-hole isn't running.

I have an Ethernet to USB adapter I can use to connect my tablet and see if I get the DNS IP I set for the Pi. Then I will know that the issue is wired vs wireless.

I am not sure how I can draw a picture. The VPN is all done locally. I am not connecting to another device before connecting to my router.

I can provide screenshots of my network adapters, my ipconfig info when i have the VPN on and when I have it off, is removing the `interface=eth0` line does not work.

**Comment by Mcat12** [comment]
Does your router strictly separate ethernet from wireless? If it doesn't, then the interface of the Pi-hole shouldn't matter as long as it's on the network because the router will route traffic to the correct place. Please send a picture of your router's DNS config, and search in there for a setting called DNS Rebind Protection and disable it if it's enabled (that stops local DNS servers from being used). If you still can't get it working, try disabling DHCP on your router and enabling it on your Pi-hole. Then reconnect your devices to the network.

**Comment by Ferkner** [comment]
There was not much in the DNS config area. just to either get the DNS automatically, or to input my own, which I did by entering the pihole IP. I did not see anything about DNS Rebind Protection. It's a modem/router combo provided by my ISP, so even the advanced view isn't that advanced. But they did leave enough there that I could change to break my Internet and Fibe TV service :)

**Comment by Mcat12** [comment]
Did you get DHCP working on your raspberry pi?

**Comment by Ferkner** [comment]
No, I didn't. When I installed it and then disabled it on my router, none of my devices could get internet access except for my desktop when the VPN was running. that's why I think it's a wired .vs wireless issue with pi-hole. I;ll find out this evening!

**Comment by Ferkner** [comment]
So I did some tweaking, and it appears that when I first installed pi-hole, I changed the DNS server on my desktop to the Pi's IP address, which is why it was working on it. After removing that, it no longer works on my desktop whether I am connected to my VPN or not.

So it is not dishing out the DNS to any machine. I connected my tablet to the network via Ethernet and it did not get the DNS from the Pi.

ifconfig info from my Pi terminal:
<img width="460" alt="pi terminal" src="https://cloud.githubusercontent.com/assets/3790980/21371731/ef114278-c6e0-11e6-969c-1cc6cde735d9.PNG">

ipconfig info from my tablet (wifi):
<img width="383" alt="surface ipconfig" src="https://cloud.githubusercontent.com/assets/3790980/21371742/015c3b22-c6e1-11e6-898f-8658ad5b4bf6.PNG">

ipconfig info from my tablet (ethernet):
<img width="366" alt="tablet ethernet" src="https://cloud.githubusercontent.com/assets/3790980/21371801/5aef1e0c-c6e1-11e6-81a6-587e9d002a6c.PNG">

DNS options on my Bell Home Hub 2000 (Sagemcom 5250):
<img width="563" alt="fibe dns" src="https://cloud.githubusercontent.com/assets/3790980/21371822/7409d788-c6e1-11e6-9895-7e32f72f9a05.PNG">

DNS options on my Bell Home Hub 2000:
<img width="476" alt="fibe dchp" src="https://cloud.githubusercontent.com/assets/3790980/21371839/84628b34-c6e1-11e6-9328-7267dddcfc5c.PNG">

So I don't have the ability to tweak things as I would with a standalone router that I buy myself. I didn't have any luck the other night with getting the DCHP server working on my Pi. I can try it again if someone knows of a reliable tutorial.

When I try to restart the dnsmasq on the Pi, I get the following error message:

> Job for dnsmasq.service failed. See 'systemctl status dnsmasq.service' and 'journalctl -xn' for details.

At this point should I just start from scratch? Reinstall Raspbian, try installing the DCHP server thing, then tryin installing pi-hole again?

**Comment by DL6ER** [comment]
Okay, your situation is now clear: No device is getting informed that is should use the Pi-hole as DNS server. Instead, all are instructed to use `192.168.2.1` which is your router.

> Job for dnsmasq.service failed. See 'systemctl status dnsmasq.service' and 'journalctl -xn' for details.

You can use the installer and chose `Update`. It will preserve your settings and correct all of your config files for you.

> I didn't have any luck the other night with getting the DCHP server working on my Pi

Wait some hours and we will release the next version of Pi-hole. You will be able to activate the new Pi-hole DHCP server with a single click! Then, go to your router and disable his DHCP server completely (Switch to OFF). Everything should work then (after you reconnect all of the clients once).

**Comment by Ferkner** [comment]
I will wait until you release the next version of pi-hole, and then update with the curl command. I assume it's being released today :)

**Comment by Tntdruid** [comment]
Does it support IPv6?

**Comment by DL6ER** [comment]
Yes, I use IPv4 + IPv6 at work. Can confirm that it is working.

**Comment by Ferkner** [comment]
So I updated pi-hole, kept the settings as before, and used the web interface to turn on the DCHP server in pi-hole, and then turned off DCHP on my Home Hub router. The result was I could connect to my network, but had no internet access

Fibe DNS still set:
<img width="574" alt="fibe dns" src="https://cloud.githubusercontent.com/assets/3790980/21413507/c91024b4-c7c5-11e6-8044-cafd89c22da9.PNG">

pi-hole DCHP on:
<img width="524" alt="pihole dchp" src="https://cloud.githubusercontent.com/assets/3790980/21413511/dd408b68-c7c5-11e6-9c82-5e7b7d34939d.PNG">

Home Hub DCHP off:
<img width="555" alt="fibe dchp" src="https://cloud.githubusercontent.com/assets/3790980/21413521/eb5ee758-c7c5-11e6-9fb7-07a6ca8ad8dc.PNG">

Pi ifconfig with pi-hole DCHP enabled:
<img width="459" alt="pi terminal" src="https://cloud.githubusercontent.com/assets/3790980/21413528/05731862-c7c6-11e6-83b5-8904602b8023.PNG">

The DNS was still not being pushed out to any of my devices. I am wondering if I should just wipe the Pi clean and start from scratch. Maybe that other DCHP server I installed is causing an issue. I don't remember which tutorial I used, so I would have no idea how to uninstall it. If that doesn't work, maybe my Home Hub router is the problem and just won't listen to what I want. Worse comes to worse, I manually enter the DNS Server on the devices that let me (my phone does not allow me to do that).

**Comment by DL6ER** [comment]
Do what it asks you to do:
```
sudo systemctl status dnsmasq.service
```

there will be some more detailed output of *why* it failed to start so that we can identify if it is the other DHCP server or something more complicated. Please also provide your `/etc/dnsmasq.d/01-pihole.conf`settings file.

**Comment by Ferkner** [comment]
Here is what got returned with that command:

```dnsmasq.service - dnsmasq - A lightweight DHCP and caching DNS server
   Loaded: loaded (/lib/systemd/system/dnsmasq.service; enabled)
  Drop-In: /run/systemd/generator/dnsmasq.service.d
           └─50-dnsmasq-$named.conf, 50-insserv.conf-$named.conf
   Active: failed (Result: exit-code) since Thu 2016-12-22 02:15:18 UTC; 10h ago
  Process: 1350 ExecStartPre=/usr/sbin/dnsmasq --test (code=exited, status=1/FAILURE)

Dec 22 02:15:18 pihole dnsmasq[1350]: dnsmasq: illegal repeated keyword at ...nf
Dec 22 02:15:18 pihole systemd[1]: dnsmasq.service: control process exited,...=1
Dec 22 02:15:18 pihole systemd[1]: Failed to start dnsmasq - A lightweight ...r.
Dec 22 02:15:18 pihole systemd[1]: Unit dnsmasq.service entered failed state.
Hint: Some lines were ellipsized, use -l to show in full.```

And attached is the 01-pihole.conf file.


[01-pihole.zip](https://github.com/pi-hole/pi-hole/files/668761/01-pihole.zip)

**Comment by DL6ER** [comment]
Okay ... follow the advice, because the shortened version is useless for me

> Hint: Some lines were ellipsized, use -l to show in full (this is not necessary on my system)

---> sudo systemctl status dnsmasq.service -l

**Comment by Ferkner** [comment]
Sorry. I'll post the full result when I get home. I was in a hurry this morning before leaving for work and didn't see the hint.

**Comment by Ferkner** [comment]
Here is the long version of the status:

`
● dnsmasq.service - dnsmasq - A lightweight DHCP and caching DNS server
   Loaded: loaded (/lib/systemd/system/dnsmasq.service; enabled)
  Drop-In: /run/systemd/generator/dnsmasq.service.d
           └─50-dnsmasq-$named.conf, 50-insserv.conf-$named.conf
   Active: failed (Result: exit-code) since Thu 2016-12-22 02:15:18 UTC; 19h ago
  Process: 1350 ExecStartPre=/usr/sbin/dnsmasq --test (code=exited, status=1/FAILURE)

Dec 22 02:15:18 pihole dnsmasq[1350]: dnsmasq: illegal repeated keyword at line 37 of /etc/dnsmasq.d/01-pihole.conf
Dec 22 02:15:18 pihole systemd[1]: dnsmasq.service: control process exited, code=exited status=1
Dec 22 02:15:18 pihole systemd[1]: Failed to start dnsmasq - A lightweight DHCP and caching DNS server.
Dec 22 02:15:18 pihole systemd[1]: Unit dnsmasq.service entered failed state.
`

**Comment by DL6ER** [comment]
Check your `/etc/dnsmasq.conf`. Either the file has to be completely empty or everything has to be commented out.

The line
> Dec 22 02:15:18 pihole dnsmasq[1350]: dnsmasq: illegal repeated keyword at line 37 of /etc/dnsmasq.d/01-pihole.conf

tells us that you have already set `cache-size=...` somewhere before.

**Comment by Ferkner** [comment]
So I made sure dnsmasq.conf was all commented out, commented out the line I had added as part of a tutorial in /etc/dnsmasq.d/01-pihole.conf and deleted the 01-pihole.conf.save file that was also there with it. Then I turned on the DCHP server on pi-hole and turned it off on my router, and now it's working for all the devices I've checked so far, including my phone.

So thank you for your great patience in helping me get this up and running. It's greatly appreciated!
